pkg/systemd/reporting.jsx

```diff
@@ -0,0 +1,450 @@
+    COMPLETED: 2,
+    ERROR: 3,
+    CANCELED: 4,
+});
+
+const PromptType = Object.freeze({
+    ASK: 0,
+    ASK_YES_NO: 1,
+    ASK_YES_NO_YESFOREVER: 2,
+    ASK_YES_NO_SAVE: 3,
+    ASK_PASSWORD: 4,
+});
+
+const ProblemState = Object.freeze({
+    REPORTABLE: 0,
+    REPORTING: 1,
+    REPORTED: 2,
+    UNREPORTABLE: 3,
+});
+
+// Avoiding using the proxy to circumvent the GLib cache.
+function get_problem_property(problem, property) {
+    return cockpit
+            .dbus("org.freedesktop.problems", { superuser: "try" })
+            .call(problem.path,
+                  "org.freedesktop.DBus.Properties",
+                  "Get",
+                  ["org.freedesktop.Problems2.Entry", property]);
+}
+
+class FAFWorkflowRow extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            problemState: ProblemState.REPORTABLE,
+            process: null,
+            reportLinks: [],
+            message: "",
+        };
+
+        this.onCancelButtonClick = this.onCancelButtonClick.bind(this);
+        this.onReportButtonClick = this.onReportButtonClick.bind(this);
+        this.updateStatusFromBus = this.updateStatusFromBus.bind(this);
+
+        this.updateStatusFromBus();
+    }
+
+    updateStatusFromBus() {
+        get_problem_property(this.props.problem, "Reports")
+                .done((reports) => {
+                    const reportLinks = [];
+                    let reported = false;
+
+                    for (const report of reports[0].v) {
+                        if (report[0] === "ABRT Server") {
+                            if ("URL" in report[1]) {
+                                reportLinks.push(report[1].URL.v.v);
+                            }
+                            reported = true;
+                        }
+                    }
+
+                    this.setState({
+                        problemState: reported ? ProblemState.REPORTED : this.state.problemState,
+                        reportLinks: reportLinks,
+                    });
+                });
+        get_problem_property(this.props.problem, "CanBeReported")
+                .done((reportable) => {
+                    if (!reportable) {
+                        this.setState({ problemState: ProblemState.UNREPORTABLE });
+                    }
+                });
+    }
+
+    onCancelButtonClick(event) {
+        this.state.process.close("canceled");
+    }
+
+    onReportButtonClick(event) {
+        if (this.state.problemState === ProblemState.REPORTING) {
+            return;
+        }
+
+        const process = cockpit.spawn(["reporter-ureport", "-d", this.props.problem.ID],
+                                      {
+                                          err: "out",
+                                          superuser: "true",
+                                      })
+                .always(() => {
+                    this.updateStatusFromBus();
+                })
```

Comment:
General style nitpick: Please avoid the unnecessary { ;} for arrow functions. These are primarily meant to be an expression, i. e. use `.always(() => this.updateStatusFromBus())`. Same issue with a lot of similar handlers in this PR.

Reply:
Sure thing.

Reply:
After a cursory glance, I don’t think any of the uses in the file actually are meant to be expressions, so, while that might be the general case in Cockpit, it doesn’t quite make sense to be returning values like that.

Reply:
Okay, I can definitely convert *some* of them, but then readability drops because of the deep nesting (which is another problem).

Reply:
Sure it's a trade-off between "lots of extra punctuation" ( { ; }) and "more stuff on the same line. This is just a nitpick though (I generally prefer more compactness and less boilerplate), but feel free to ignore. This is by far the least important review comment here. :grinning: